File: src/y2022/day11.rs
//! <https://adventofcode.com/2022/day/11>
use {
    crate::{
        framework::{AdventOfCode, ParseError, aoc},
        parser::parse_usize,
    },
    winnow::{
        ModalResult, Parser,
        ascii::{newline, space1},
        combinator::{alt, separated, seq},
        token::one_of,
    },
};

#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
struct Monkey {
    id: usize,
    items: Vec<usize>,
    operation: (bool, Option<usize>),
    test: usize,
    test_then: usize,
    test_else: usize,
    num_inspect: usize,
}

impl Monkey {
    fn update(&mut self, thrown: &mut [Vec<usize>]) {
        for i in self.items.iter() {
            self.num_inspect += 1;
            let mut j = match self.operation {
                (false, None) => i + i,
                (false, Some(k)) => i + k,
                (true, None) => i * i,
                (true, Some(k)) => i * k,
            };
            j /= 3;
            thrown[if j % self.test == 0 {
                self.test_then
            } else {
                self.test_else
            }]
            .push(j);
        }
        self.items.clear();
    }
    fn update2(&mut self, thrown: &mut [Vec<usize>], cd: usize) {
        for i in self.items.iter() {
            self.num_inspect += 1;
            let j = match self.operation {
                (false, None) => i + i,
                (false, Some(k)) => i + k,
                (true, None) => i * i,
                (true, Some(k)) => i * k,
            };
            thrown[if j % self.test == 0 {
                self.test_then
            } else {
                self.test_else
            }]
            .push(j % cd);
        }
        self.items.clear();
    }
}

#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Puzzle {
    line: Vec<Monkey>,
}

fn parse_operation_target(s: &mut &str) -> ModalResult<Option<usize>> {
    alt(("old".map(|_a| None), parse_usize.map(Some))).parse_next(s)
}

fn parse_block(s: &mut &str) -> ModalResult<Monkey> {
    seq!(
        _: "Monkey ", parse_usize, _: ":",
        _: "\n  Starting items: ", separated(1.., parse_usize, ", "),
        _: "\n  Operation: new = old ", one_of(['*', '+']), _: space1, parse_operation_target,
        _: "\n  Test: divisible by ", parse_usize,
        _: "\n    If true: throw to monkey ", parse_usize,
        _: "\n    If false: throw to monkey ", parse_usize,
        _: newline
    )
    .map(|(id, items, op, num1, test, test_then, test_else)| Monkey {
        id,
        items,
        operation: (op == '*', num1),
        test,
        test_then,
        test_else,
        ..Default::default()
    })
    .parse_next(s)
}

fn parse(s: &mut &str) -> ModalResult<Vec<Monkey>> {
    separated(1.., parse_block, newline).parse_next(s)
}

#[aoc(2022, 11)]
impl AdventOfCode for Puzzle {
    fn parse(&mut self, mut input: &str) -> Result<(), ParseError> {
        self.line = parse(&mut input)?;
        Self::parsed()
    }
    fn part1(&mut self) -> Self::Output1 {
        let mut tmp = vec![Vec::new(); self.line.len()];
        for _ in 0..20 {
            for i in 0..self.line.len() {
                let m = &mut self.line[i];
                m.update(&mut tmp);
                self.thrown(&mut tmp);
            }
            // println!(
            //     "{:?}",
            //     self.line.iter().map(|m| &m.items).collect::<Vec<_>>()
            // );
        }
        // dbg!(self
        //     .line
        //     .iter()
        //     .map(|m| m.num_inspect)
        //     .collect::<Vec<usize>>());
        let mut v = self
            .line
            .iter()
            .map(|m| m.num_inspect)
            .collect::<Vec<usize>>();
        v.sort();
        v.reverse();
        v[0] * v[1]
    }
    fn part2(&mut self) -> Self::Output2 {
        let cd = self.line.iter().map(|m| m.test).product();
        let mut tmp = vec![Vec::new(); self.line.len()];
        for _ in 0..10000 {
            for i in 0..self.line.len() {
                let m = &mut self.line[i];
                m.update2(&mut tmp, cd);
                self.thrown(&mut tmp);
            }
        }
        // dbg!(self
        //     .line
        //     .iter()
        //     .map(|m| m.num_inspect)
        //     .collect::<Vec<usize>>());
        let mut v = self
            .line
            .iter()
            .map(|m| m.num_inspect)
            .collect::<Vec<usize>>();
        v.sort();
        v.reverse();
        v[0] * v[1]
    }
}

impl Puzzle {
    fn thrown(&mut self, items: &mut [Vec<usize>]) {
        for (i, v) in items.iter_mut().enumerate() {
            if !v.is_empty() {
                let mut w = Vec::new();
                std::mem::swap(v, &mut w);
                self.line[i].items.append(&mut w);
            }
        }
    }
}

File: Swift/Sources/Y2022/day11.swift
//
//  day11.swift
//  aoc
//
import Parsing

@DebugDescription
private class Monkey {
    let id: Int
    var items: [Int]
    let operation: (Bool, Int?)
    let test: Int
    let test_then: Int
    let test_else: Int
    var num_inspect: Int
    init(
        id: Int,
        items: [Int],
        operation: (Bool, Int?),
        test: Int,
        test_then: Int,
        test_else: Int,
    ) {
        self.id = id
        self.items = items
        self.operation = operation
        self.test = test
        self.test_then = test_then
        self.test_else = test_else
        self.num_inspect = 0
    }
    func copy() -> Monkey {
        return Monkey(
            id: self.id,
            items: self.items,
            operation: self.operation,
            test: self.test,
            test_then: self.test_then,
            test_else: self.test_else
        )
    }
    var debugDescription: String {
        "\(id)"
    }
    func update(thrown: inout [[Int]]) {
        for i in self.items {
            self.num_inspect += 1
            var j =
                switch self.operation {
                case (false, let k): i + (k ?? 1)
                case (true, let k): i * (k ?? i)
                }
            j /= 3
            thrown[(j % self.test == 0) ? self.test_then : self.test_else].append(j)
        }
        self.items.removeAll()
    }
    func update2(thrown: inout [[Int]], dividedBy d: Int) {
        for i in self.items {
            self.num_inspect += 1
            let j =
                switch self.operation {
                case (false, let k): i + (k ?? 1)
                case (true, let k): i * (k ?? i)
                }
            thrown[(j % self.test == 0) ? self.test_then : self.test_else].append(j % d)
        }
        self.items.removeAll()
    }
}

private func update(monkeys: inout [Monkey], state: inout [[Int]]) {
    for (i, v) in state.enumerated() {
        if !v.isEmpty {
            state[i] = []
            monkeys[i].items.append(contentsOf: v)
        }
    }
}

private func part1(monkeys: [Monkey]) -> Int {
    var monkeys = monkeys.map { $0.copy() }
    var state: [[Int]] = Array(repeating: Array(), count: monkeys.count)
    for _ in 1...20 {
        for monkey in monkeys {
            monkey.update(thrown: &state)
            update(monkeys: &monkeys, state: &state)
        }
    }
    let v: [Int] = monkeys.map(\.num_inspect).sorted().reversed()
    return v[0] * v[1]
}

private func part2(monkeys: [Monkey]) -> Int {
    var monkeys = monkeys.map { $0.copy() }
    let cd: Int = monkeys.reduce(1) { $0 * $1.test }
    var state: [[Int]] = Array(repeating: Array(), count: monkeys.count)
    for _ in 1...10000 {
        for monkey in monkeys {
            monkey.update2(thrown: &state, dividedBy: cd)
            update(monkeys: &monkeys, state: &state)
        }
    }
    let v: [Int] = monkeys.map(\.num_inspect).sorted().reversed()
    return v[0] * v[1]
}

public func day11(_ data: String) {
    let operation_target: some Parser<Substring, Int?> = OneOf {
        Parse { "old" }.map { _ in nil as Int? }
        Int.parser().map { $0 as Int? }
    }
    let monkey: some Parser<Substring, Monkey> = Parse {
        "Monkey "
        Int.parser()
        ":\n  Starting items: "
        Many {
            Int.parser()
        } separator: {
            ", "
        }
        "\n  Operation: new = old "
        OneOf {
            "* ".map { true }
            "+ ".map { false }
        }
        operation_target
        "\n  Test: divisible by "
        Int.parser()
        "\n    If true: throw to monkey "
        Int.parser()
        "\n    If false: throw to monkey "
        Int.parser()
        "\n"
    }.map { id, items, isMul, num1, test, test_then, test_else in
        Monkey(
            id: id,
            items: items,
            operation: (isMul, num1),
            test: test,
            test_then: test_then,
            test_else: test_else,
        )
    }
    let parser: some Parser<Substring, [Monkey]> = Many {
        monkey
    } separator: {
        "\n"
    }
    do {
        let input = try parser.parse(data)
        let sum1 = part1(monkeys: input)
        let sum2 = part2(monkeys: input)
        print("Part 1: \(sum1)")
        print("Part 2: \(sum2)")
    } catch {
        print(error)
        fatalError()
    }
}
